lib/time/time.go

```diff
@@ -0,0 +1,548 @@
+}
+
+// Binary implements binary operators, which satisfies the starlark.HasBinary
+// interface
+//    time + duration = time
+//    time - duration = time
+//    time - time = duration
+func (t Time) Binary(op syntax.Token, yV starlark.Value, side starlark.Side) (starlark.Value, error) {
+	x := time.Time(t)
+
+	switch op {
+	case syntax.PLUS:
+		switch y := yV.(type) {
+		case Duration:
+			return Time(x.Add(time.Duration(y))), nil
+		case Time:
+			return nil, fmt.Errorf("cannot add %s to %s", t.Type(), yV.Type())
+		}
+	case syntax.MINUS:
+		switch y := yV.(type) {
+		case Duration:
+			return Time(x.Add(time.Duration(-y))), nil
+		case Time:
+			// time - time = duration
+			if side == starlark.Left {
+				return Duration(x.Sub(time.Time(y))), nil
+			} else {
+				return Duration(time.Time(y).Sub(x)), nil
+			}
+		}
+	}
+
+	return nil, nil
+}
+
+var timeMethods = map[string]builtinMethod{
+	"in_location": timeIn,
+	"format":      timeFormat,
+}
+
+func timeFormat(fnname string, recV starlark.Value, args starlark.Tuple, kwargs []starlark.Tuple) (starlark.Value, error) {
+	var x starlark.String
+	if err := starlark.UnpackArgs("format", args, kwargs, "x", &x); err != nil {
+		return nil, err
+	}
+
+	recv := time.Time(recV.(Time))
+	return starlark.String(recv.Format(string(x))), nil
+}
+
+func timeIn(fnname string, recV starlark.Value, args starlark.Tuple, kwargs []starlark.Tuple) (starlark.Value, error) {
+	var x starlark.String
+	if err := starlark.UnpackArgs("in_location", args, kwargs, "x", &x); err != nil {
+		return nil, err
+	}
+	loc, err := time.LoadLocation(string(x))
+	if err != nil {
+		return nil, err
+	}
+
+	recv := time.Time(recV.(Time))
+	return Time(recv.In(loc)), nil
+}
+
+type builtinMethod func(fnname string, recv starlark.Value, args starlark.Tuple, kwargs []starlark.Tuple) (starlark.Value, error)
+
+func builtinAttr(recv starlark.Value, name string, methods map[string]builtinMethod) (starlark.Value, error) {
+	method := methods[name]
+	if method == nil {
+		return nil, nil // no such method
+	}
+
+	// Allocate a closure over 'method'.
+	impl := func(thread *starlark.Thread, b *starlark.Builtin, args starlark.Tuple, kwargs []starlark.Tuple) (starlark.Value, error) {
+		return method(b.Name(), b.Receiver(), args, kwargs)
+	}
+	return starlark.NewBuiltin(name, impl).BindReceiver(recv), nil
+}
+
+func builtinAttrNames(methods map[string]builtinMethod) []string {
+	names := make([]string, 0, len(methods))
+	for name := range methods {
+		names = append(names, name)
+	}
+	sort.Strings(names)
+	return names
+}
+
+// threeway interprets a three-way comparison value cmp (-1, 0, +1)
+// as a boolean comparison (e.g. x < y).
+func threeway(op syntax.Token, cmp int) bool {
+	switch op {
+	case syntax.EQL:
+		return cmp == 0
+	case syntax.NEQ:
+		return cmp != 0
+	case syntax.LE:
+		return cmp <= 0
+	case syntax.LT:
+		return cmp < 0
+	case syntax.GE:
+		return cmp >= 0
+	case syntax.GT:
+		return cmp > 0
+	}
+	panic(op)
+}
```

Comment:
You use this function exactly once in `CompareSameType()`, so maybe you should move the implementation just there.

Reply:
It's fine to keep it separate, as it crops up a lot and this makes clear that it's a simply copy.

(The design of CompareSameType is overcomplicated; see  3b7e02ecde8b80808c06c8284ce9f85c101fcebd for context.)

Reply:
Ok but then please make this an infrastructure and only ask the new types to compute a comparison between `a` and `b` by returning `-1`, `0` or `+1` indicting if `a<b`, `a==b` or `a>b` and do the `threeway` logic elsewhere? :-)

Reply:
Yes, we should do that. (Or perhaps define a new, simpler CompareSameType2 interface and make the interpreter test it first.) I've filed https://github.com/google/starlark-go/issues/356.